web/base_viewer.js

```diff
@@ -63,8 +78,13 @@ function PDFPageViewBuffer(size) {
       data.shift().destroy();
     }
   };
-  this.resize = function(newSize) {
+  this.resize = function(newSize, setToKeep) {
     size = newSize;
+    if (setToKeep) {
+      moveToEndOfArray(data, function(page) {
+        return setToKeep.has(page.id);
+      });
```

Comment:
That ‘side effect’ is exactly the intent (and I'd hoped that the name of `moveToEndOfArray` would make it clear that it is a mutator).

This is another change to all scrolling modes that I think is justified, even though it was unlikely to cause visible problems in the original code. The buffer is a simple queue, in which pages are destroyed when they reach the end of the queue. But the queue order is determined by the order in which the pages were first drawn, which is not necessarily the order in which they should be forgotten if the user is scrolling both backward and forward. For example, if you jump to page 20, and then scroll back several pages such that 15 gets drawn, and then start scrolling forward from there, page 20 will eventually be destroyed before pages 15–19, leaving a ‘hole’. For buffers that are large relative to the number of visible pages, this is not going to be a very visible problem. But with wrapped scrolling in particular, the number of visible pages can increase to the point where, without this change, it's possible to see already rendered pages being destroyed and then re-rendered! (This happens even though the buffer size is greater than the number of visible pages, because the buffer size changes when switching to a scrolling mode where more pages can be visible.)

This change not only prevents any pages that are currently visible from being wastefully destroyed, it moves them to the back of the queue so that they'll last longer than any pages that are currently not visible, reducing the likelihood of ‘holes’.

Reply:
Makes sense.